### DISCUSSION TITLE
Perspectives from cryptography

### DISCUSSION THREAD

**Discussion by oskgo** [blainehansen/magmide#17] [Ideas]
#### Why cryptography is relevant for Magmide

1. The field is already doing formal verification of imperative "programs" on a sizable scale. Cryptographers care about their algorithms being designed and implemented correctly.
2. Cryptographic algorithms are one of the more accessible use cases of Magmide since they can be pretty small.
3. Cryptographers, for the most part, prefer imperative over functional programming. One reason why is that having input independent run-time is easier with imperative than functional programming, another is that imperative languages tend to be faster.

#### State of formal verification in cryptography

Formal verification is mostly used in two ways, firstly it's used to show that an implementation of a cryptographic algorithm matches some specification. This is one of the main use-cases of F* in project Everest (check out [Idris](https://www.idris-lang.org/) btw, it's very similar to F*).

Secondly it's used to prove that a cryptographic algorithm is "correct" (e.g. decrypting an encrypted message gives back the original) and that it is "secure" (for some slightly dishonest definition of secure). In order to prove "security" and often also "correctness" we have to be able to reason about probabilistic programs. This is out of scope for F* and Idris. Coq and Isabelle both have libraries for reasoning about probabilistic algorithms ([FCF](https://github.com/adampetcher/fcf) and [CryptHOL](https://www.isa-afp.org/entries/CryptHOL.html)), but these will likely never be used outside cryptography. There is also a standalone proof assistant for reasoning about cryptography, [EasyCrypt](https://github.com/EasyCrypt/easycrypt), but this will also never be used outside cryptography. This is where Magmide can make a difference and provide something that is of interest to a broader community and thus has a larger ceiling on development time, library count and time spent "looking for bugs".

#### What Magmide can learn from cryptography

[Jasmin](https://github.com/jasmin-lang/jasmin) is a c-like programming language being developed for doing cryptography. Its semantics are defined in Coq and it exports to EasyCrypt for formal verification. The compiler is verified in Coq. I think that it does a lot of the things you are interested in doing for Host Magmide, though it isn't built for uses outside cryptography.

#### What Magmide has to do to have first-class support for reasoning about cryptography

Most features, like asserting that a function has runtime independent of secret inputs can be added far into development, but I believe that support for reasoning about probabilistic programs has to be there from early on (before there would be resistance to changing how programs are modeled), otherwise there's the risk that the language is split into two parts, the probabilistic and the deterministic ones, with lacking compatibility.

The usual way to do this is to replace the option monad in Hoare logic with a distribution monad ([this (currently incomplete) book](https://software.imdea.org/~gbarthe/__introrelver.pdf) shows some of the theory).

**Comment by blainehansen** [comment]
I'll absolutely look into this! I don't know almost anything about cryptography, and honestly I'm surprised that F* doesn't handle runtime dependence (I'm sure you've heard of [vale](https://www.microsoft.com/en-us/research/publication/vale-verifying-high-performance-cryptographic-assembly-code/)? is handling side channels equivalent to ensuring input independent runtime?), but yes cryptography is absolutely intended to be a well-supported use case. If you have more resources about different ways to allow these use cases, please share them!
I wish I was making faster progress, but I've been having trouble finding the time and space and support to do so.

Thank you for pointing this out! As the design and implementation progresses I'll keep this in mind :)

**Reply by oskgo** [reply]
Handling side channels in general is a futile effort. Ensuring (secret) input independent runtime is a mitigation against timing based side channels. I'm sure that that's what Microsoft means though.

I don't know if F* handles this, it's more of a point against functional programming in general. In functional programs you usually have a lot less control over the output machine code. Ensuring input independent runtime for assembly amounts to avoiding a few no-go instructions (division on some archs) on data dependent on secrets ~~and doing some logic to check that same instructions are being used on every branch that depends on a secret~~.

**Reply by programmerjake** [reply]
iirc branching on secret values is a major no-no, because the branch predictor and stuff likes to make it take different amounts of time for a taken vs. not taken branch.

**Reply by oskgo** [reply]
> iirc branching on secret values is a major no-no, because the branch predictor and stuff likes to make it take different amounts of time for a taken vs. not taken branch.

You're right, thanks. My inexperience with the implementation side is showing.

**Reply by klinvill** [reply]
Just to add some additional details about F*, generally F* is expressive enough to allow you to represent additional logics (e.g. probabilistic Relational Hoare Logic, https://arxiv.org/abs/1703.00055, like EasyCrypt uses for relational proofs about probabilities) and languages (e.g. Vale for x64 assembly, https://eprint.iacr.org/2019/542.pdf, and Low* for C) within F* itself. You can then use these subsets of the logics and languages to write programs that can be extracted to those languages (via something like KaRaMeL, nee KReMLin, https://github.com/FStarLang/karamel) (and compiled using a verified compiler) for actual execution. Through the use of its monadic effect system, you can also represent stateful imperative computations with a generally convenient syntax (example below). You can generally do the same thing with functional languages, it just starts adding layers of additional implementation vs. something like Dafny that lets you just write imperative code.

So the F* vale crypto work does check to ensure the assembly is secret input independent but it requires a lot of additional support added to F* and requires you to essentially write your own assembly/C code rather than writing arbitrary F* code. 

There's generally a tradeoff here between designing a general purpose language for writing and verifying programs (which can pretty much support any use case but can require a lot of extra implementation to embed a logic or language) vs. designing a domain-specific language/framework like EasyCrypt that is much easier to use out-of-the-box but comes with restrictions that prevent it from really being useful for other domains.

Using monadic effects in F* looks something like the following:
```fstar
// StInt is a monadic effect that keeps track of a stateful integer
let incr_by (value: int) : StInt unit =
  // Here `ST_INT?.get ()` gets the current state 
  let state = ST_INT?.get () in
  // Like many languages targetted towards PL folks, a single quote (pronounced prime) in a variable name generally denotes that it is the transformed or updated version of the unprimed variable, rather than opening a string like in most languages targetted for devs
  let state' = state + value in
  // Here `ST_INT?.get ()` writes a new value to the current state
  ST_INT?.put state';
  ()
```

You can then use functions with this effect in an imperative like manner:
```fstar
let incr_thrice (value: int) : StInt unit =
  incr_by value;
  incr_by value;
  incr_by value
```
